### PR TITLE
Adding MLX backend.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Adds MLX backend.
MLX already has native safetensors support, but this was trivial to add.

Fixes # (issue) or description of the problem this PR solves.

### CODE REVIEW

**Comment by Narsil** [comment]
We can safely ignore the failure which is entirely unrelated to the PR (more recent clippy detects an issue with PyO3 itself, which will most likely be fixed, by simply upgrading PyO3)

**Review by Vaibhavs10 — COMMENTED** [review]
Dope! Thanks for putting this up! 🚀 Conceptually, this looks good to me.

That said, I'm not an expert in MLX so would appreciate if Awni could do a deeper review.

One qq: Can we potentially also test for persisting and loading quantised models as well? we don't need to be exhaustive here, but just testing for 4-bit should be okay IMO.

File: bindings/python/py_src/safetensors/mlx.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    save_file(tensors, "model.safetensors")
+    ```
+    """
+    np_tensors = _mx2np(tensors)
+    return numpy.save_file(np_tensors, filename, metadata=metadata)
+
+
+def load(data: bytes) -> Dict[str, mx.array]:
+    """
+    Loads a safetensors file into flax format from pure bytes.
```
Comment (pcuenca) [inline bindings/python/py_src/safetensors/mlx.py:76]:
```suggestion
    Loads a safetensors file into MLX format from pure bytes.
```

File: bindings/python/py_src/safetensors/mlx.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        data (`bytes`):
+            The content of a safetensors file
+
+    Returns:
+        `Dict[str, mx.array]`: dictionary that contains name as key, value as `mx.array` on cpu
+
+    Example:
+
+    ```python
+    from safetensors.flax import load
```
Comment (pcuenca) [inline bindings/python/py_src/safetensors/mlx.py:88]:
```suggestion
    from safetensors.mlx import load
```

File: bindings/python/py_src/safetensors/mlx.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    loaded = load(data)
+    ```
+    """
+    flat = numpy.load(data)
+    return _np2mx(flat)
+
+
+def load_file(filename: Union[str, os.PathLike]) -> Dict[str, mx.array]:
+    """
+    Loads a safetensors file into flax format.
```
Comment (pcuenca) [inline bindings/python/py_src/safetensors/mlx.py:103]:
```suggestion
    Loads a safetensors file into MLX format.
```

File: bindings/python/py_src/safetensors/mlx.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    Args:
+        filename (`str`, or `os.PathLike`)):
+            The name of the file which contains the tensors
+        device (`Dict[str, any]`, *optional*, defaults to `cpu`):
+            The device where the tensors need to be located after load.
+            available options are all regular flax device locations
```
Comment (pcuenca) [inline bindings/python/py_src/safetensors/mlx.py:110]:
```suggestion
```

**Comment by Narsil** [comment]
Quantized is u8, therefore it should work out of the box (not sure how mlx persists quantized information in npx/npz, i'm guessing it's not saved)

File: bindings/python/py_src/safetensors/mlx.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+def load(data: bytes) -> Dict[str, mx.array]:
+    """
+    Loads a safetensors file into flax format from pure bytes.
+
+    Args:
+        data (`bytes`):
+            The content of a safetensors file
+
+    Returns:
+        `Dict[str, mx.array]`: dictionary that contains name as key, value as `mx.array` on cpu
```
Comment (pcuenca) [inline bindings/python/py_src/safetensors/mlx.py:83]:
```suggestion
        `Dict[str, mx.array]`: dictionary that contains name as key, value as `mx.array`
```

File: bindings/python/py_src/safetensors/mlx.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    return result
+
+
+def _np2mx(numpy_dict: Dict[str, np.ndarray]) -> Dict[str, mx.array]:
+    for k, v in numpy_dict.items():
+        numpy_dict[k] = mx.array(v)
+    return numpy_dict
+
+
+def _mx2np(mx_dict: Dict[str, mx.array]) -> Dict[str, np.array]:
```
Comment (pcuenca) [inline bindings/python/py_src/safetensors/mlx.py:137]: Does this handle `bfloat16`?

File: bindings/python/tests/test_mlx_comparison.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+if platform.system() == "Darwin":
+    # This platform is not supported, we don't want to crash on import
+    # This test will be skipped anyway.
+    import mlx.core as mx
+    from safetensors import safe_open
+    from safetensors.mlx import load_file, save_file
+
+
+# Jax doesn't not exist on Windows
```
Comment (pcuenca) [inline bindings/python/tests/test_mlx_comparison.py:13]:
```suggestion
# MLX only exists on Mac
```

File: bindings/python/py_src/safetensors/mlx.py (hunk 6 above)
Comment (pcuenca) [reply bindings/python/py_src/safetensors/mlx.py:137]: Oh, I see in the tests below that it does :)

**Review by pcuenca — APPROVED** [review]
Awesome! 🔥

File: bindings/python/py_src/safetensors/mlx.py (hunk 6 above)
Comment (Narsil) [reply bindings/python/py_src/safetensors/mlx.py:137]: Yup, using the same flax trick which is just a special named dtype.

**Review by awni — APPROVED** [review]
LGTM! Thanks for adding that!

I think using your built-in `numpy.save` is probably a good call from a maintenance standpoint.

But just a thought that this will involve a copy (on the load side as well). Currently we do a copy to get in and out of Numpy.

**Comment by Narsil** [comment]
> But just a thought that this will involve a copy (on the load side as well). Currently we do a copy to get in and out of Numpy.


Good to know. The only thing really needed is `frombuffer` equivalent in order to get loading partial tensors working.
On the rust side, I use the slice information given by the user to create. a CPU buffer of the correct data within the tensor, and use `frombuffer`  on given framework in order to send the correct data where it belongs.

If there's a copy from numpy-> MLX, that means that currently there are 2 copies created.

File -> Slice (local buffer inside rust) -> numpy.from_buffer (should be zero copy) -> mlx (new copy).

Is that correct ?

**Comment by awni** [comment]
> If there's a copy from numpy-> MLX, that means that currently there are 2 copies created.

Yes that's correct. We can't avoid the copy from NumPy because we have to use specifically allocated memory to make it available to both the CPU and GPU.

**Pull request merged**